spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/CassandraRow.scala

```diff
@@ -72,12 +72,65 @@ import com.datastax.driver.core.Row
   *   - java.sql.Date
   *   - org.joda.time.DateTime
   */
-final class CassandraRow(val columnNames: IndexedSeq[String], val columnValues: IndexedSeq[AnyRef])
+final class CassandraRow(val metaData:CassandraRowMetaData, val columnValues: IndexedSeq[AnyRef])
   extends ScalaGettableData with Serializable {
 
+  /**
+    * the consturctor is for testing and backward compatibility only.
+    * Use default constructor with shared metadata for memory saving and performance.
+    *
+    * @param columnNames
+    * @param columnValues
+    */
+  def this (columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef]) =
+    this(CassandraRowMetaData.fromColumnNames(columnNames), columnValues)
+
   override def toString = "CassandraRow" + dataAsString
 }
 
+case class CassandraRowMetaData(columnNames: IndexedSeq[String], resultSetColumnNames:Option[IndexedSeq[String]] = None, @transient codecs: Option[IndexedSeq[TypeCodec[AnyRef]]] = None) {
```

Comment:
I think because of nulling the codecs field, it would be much safer to make them `private`. They seem to be needed only for constructing `CassandraRows` and nothing more